pkg/controller/attach_detach.go

```diff
@@ -0,0 +1,586 @@
+	return r.attachmentProvisioner.UnpublishVolume(ctx, volumeID, node)
+}
+
+func (r *ReconcileAttachDetach) Recover(ctx context.Context) error {
+	klog.Info("Recovering AzVolumeAttachment CRIs...")
+	// try to recover states
+	var syncedVolumeAttachments, volumesToSync map[string]bool
+	for i := 0; i < maxRetry; i++ {
+		var retry bool
+		var err error
+
+		retry, syncedVolumeAttachments, volumesToSync, err = r.syncAll(ctx, syncedVolumeAttachments, volumesToSync)
+		if err != nil {
+			klog.Warningf("failed to complete initial AzVolumeAttachment sync: %v", err)
+		}
+		if !retry {
+			break
+		}
+	}
+
+	return nil
+}
+
+// ManageAttachmentsForVolume will be running on a separate channel
+func (r *ReconcileAttachDetach) syncAll(ctx context.Context, syncedVolumeAttachments map[string]bool, volumesToSync map[string]bool) (bool, map[string]bool, map[string]bool, error) {
+	// Get all volumeAttachments
+	volumeAttachments, err := r.kubeClient.StorageV1().VolumeAttachments().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return true, syncedVolumeAttachments, volumesToSync, err
+	}
+
+	if syncedVolumeAttachments == nil {
+		syncedVolumeAttachments = map[string]bool{}
+	}
+	if volumesToSync == nil {
+		volumesToSync = map[string]bool{}
+	}
+
+	// Loop through volumeAttachments and create Primary AzVolumeAttachments in correspondence
+	for _, volumeAttachment := range volumeAttachments.Items {
+		// skip if sync has been completed volumeAttachment
+		if syncedVolumeAttachments[volumeAttachment.Name] {
+			continue
+		}
+		if volumeAttachment.Spec.Attacher == azureutils.DriverName {
+			volumeName := volumeAttachment.Spec.Source.PersistentVolumeName
+			if volumeName == nil {
+				continue
+			}
+			// get PV and retrieve diskName
+			pv, err := r.kubeClient.CoreV1().PersistentVolumes().Get(ctx, *volumeName, metav1.GetOptions{})
+			if err != nil {
+				klog.Errorf("failed to get PV (%s): %v", *volumeName, err)
+				return true, syncedVolumeAttachments, volumesToSync, err
+			}
+
+			if pv.Spec.CSI == nil || pv.Spec.CSI.Driver != azureutils.DriverName {
+				continue
+			}
+			volumesToSync[pv.Spec.CSI.VolumeHandle] = true
+
+			diskName, err := azureutils.GetDiskNameFromAzureManagedDiskURI(pv.Spec.CSI.VolumeHandle)
+			if err != nil {
+				klog.Warningf("failed to extract disk name from volumehandle (%s): %v", pv.Spec.CSI.VolumeHandle, err)
+				delete(volumesToSync, pv.Spec.CSI.VolumeHandle)
+				continue
+			}
+			nodeName := volumeAttachment.Spec.NodeName
+			azVolumeAttachmentName := azureutils.GetAzVolumeAttachmentName(diskName, nodeName)
+
+			// check if the CRI exists already
+			azVolumeAttachment, err := azureutils.GetAzVolumeAttachment(ctx, r.client, r.azVolumeClient, azVolumeAttachmentName, r.namespace, false)
+			klog.Infof("Recovering AzVolumeAttachment(%s)", azVolumeAttachmentName)
+			// if CRI already exists, append finalizer to it
+			if err == nil {
```

Comment:
Should we also check if the finalizers already exist on the object? In that case we can skip appending the finalizer.

Reply:
Nvm, looks like that's already being checked in InitializeMeta

Reply:
We already do that in `initializeMeta`. So, if there already are necessary finalizers and labels present in CRI, the update operation would essentially be a null op.
https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/2260a7332293f384c962089db81bbdb0940c79cf/pkg/controller/attach_detach.go#L243-L245

Reply:
Ah I was too late with my comment :D